presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -167,23 +174,40 @@ public GroupedAccumulator createGroupedAccumulator()
                 argumentTypes.add(sourceTypes.get(input));
             }
 
-            accumulator = new DistinctingGroupedAccumulator(accumulator, argumentTypes, inputChannels, maskChannel, session, joinCompiler);
+            accumulator = new SpillableDistinctingGroupedAccumulator(accumulator, argumentTypes, sourceTypes, inputChannels, maskChannel, session, joinCompiler);
         }
         else {
             accumulator = instantiateGroupedAccumulator(inputChannels, maskChannel);
         }
 
-        if (orderByChannels.isEmpty()) {
+        if (!hasOrderBy()) {
             return accumulator;
         }
 
-        return new OrderingGroupedAccumulator(accumulator, sourceTypes, orderByChannels, orderings, pagesIndexFactory);
+        return new SpillableOrderingGroupedAccumulator(accumulator, sourceTypes, orderByChannels, orderings, pagesIndexFactory);
```

Comment:
All of these need be conditionally created based on if spilling enabled/disabled + if we are using HashAggregationOperator. Not sure how to do this cleanly so left this as the default option for now.